Extending GroupItemMetadataProvider

arashdalir opened the pull request:
`Slick.Data.GroupItemMetadataProvider` now returns its `options`, making it possible to extend the default functionality.

Following code shows an example:

```javascript
window.GroupItemMetadataProvider = function (options){
	var _provider = Slick.Data.GroupItemMetadataProvider(options);

	options = _provider.getOptions();

	var defaultGroupCellFormatter = options.groupFormatter;

	options.groupFormatter = groupCellFormatter;

	function groupCellFormatter(row, cell, value, columnDef, item, grid)
	{
		if(typeof item.title !== typeof {})
		{
			item.title = {
				text: item.title,
				toolTip: null,
				addClasses: null
			}
		}

		if(!item.title.alreadyFormatted)
		{
			item.title.alreadyFormatted = true;

			item.title.text = defaultGroupCellFormatter.call(
				_provider,
				row,
				cell,
				value,
				columnDef,
				$.extend(true, {}, item, {title: item.title.text}),
				grid
			);
		}
		return item.title;
	}

	return {
		"init": _provider.init,
		"destroy": _provider.destroy,
		"getGroupRowMetadata": _provider.getGroupRowMetadata,
		"getTotalsRowMetadata": _provider.getTotalsRowMetadata
	};
}
```

ghiscoding commented:
If you want to change some of the options afterward, why not expose a `setOption()` method like I've done in the [Grid Menu](https://github.com/6pac/SlickGrid/blob/master/controls/slick.gridmenu.js#L149)?

arashdalir commented:
well, this is principally what I have done, isn't it? ;)

I wanted to be able to use the predefined functions defined in `Slick.Data.GroupItemMetadataProvider` like the default function for creating the group-cell (`defaultGroupCellFormatter`). but as the function is private, it's not possible to use it outside the scope of the class.  
also, it is not possible to access the `_defaults`, which are used during the creation of group-cell.

By exposing the `options` using `getOptions` and `setOptions`, one can now create an object of type `Slick.Data.GroupItemMetadataProvider` and access its `_defaults` (including `defaultGroupCellFormatter`) by passing an empty `options` to the `constructor`. 

as seen in the example I posted [here](https://github.com/6pac/SlickGrid/pull/466#issue-378303155), the extension allows other classes to imitate the default behavior and extend that for other uses. 

**As background to the issue**: the return value from the `formatter` defined while calling `Slick.Grid.setGrouping` is not compatible with that of `columnDef.formatter`. it can return a DOM object as string and that's it. The reason behind that is, that the `defaultGroupCellFormatter` works only with `item.title`, which the mentioned DOM string.   
I want to extend this in order to be able to use `addClasses` and `toolTip` properties, to conditionally add new styles or tooltips to entire group-cell rows. 
My example can partially do that. `addClasses` is used, but for a reason I have not been able to figure out yet, the `toolTip` value is ignored. it uses the `defaultGroupCellFormatter` to create the proper DOM string, then encapsulates that in the proper object which has following format (to be exact, it returns the same object as returned by `Slick.Grid.setGrouping([{formatter:function(){}}])`, whose `text` property is passed to the `defaultGroupCellFormatter`):
```JSON
{
    text: '',
    toolTip: '',
    addClasses:''
    ...
}
```

ghiscoding commented:
> well, this is principally what I have done, isn't it? ;)

Oh dang sorry I was actually looking at the code you've put in the comment instead of the real code in the PR, it's morning and I didn't have my coffee yet lol
I think your PR is fine by me.

Can you do a print screen of the result in the grid of the issue you're describing? Just curious and can't really visualize it, probably because I didn't have coffee yet again haha

arashdalir commented:
> Oh dang sorry I was actually looking at the code you've put in the comment instead of the real code in the PR, it's morning and I didn't have my coffee yet lol
> I think your PR is fine by me.

Sorry, my mistake, I had a "small" problem with the commit message. but no worries, mate! 

> Can you do a print screen of the result in the grid of the issue you're describing? Just curious and can't really visualize it, probably because I didn't have coffee yet again haha

Updated the answer a few times and added a small screenshot (cannot post full rows, as the data is a bit sensitive! :D). The `toolTip` (which I can confirm is correctly returned by my function - screenshot below) should read "Source - Bahrain - Bahrain Tel..." but it is just the `.text()` value or the row instead, which looks like in the screenshot provided in previous comment. the `addClasses`: `source-diff` value is applied (hence, the red bg for 3rd row), but the `toolTip` comes from somewhere else. 

![image](https://user-images.githubusercontent.com/2896170/75045639-f260af80-54c3-11ea-8a89-28fc48dd8f4c.png)

arashdalir commented:
if you need more details or a bit of help imagining the issue,  I can arrange a short skype session in order to share my screen with you, but unfortunately cannot post complete examples, as we have extended the `slick-grid` functionality and an example will contain too many complicated files...

ghiscoding commented:
I'm ok with your PR if it fixes your issue. I was just trying to see if I missed something in the previous PR #300 for the tooltip but I don't have time to look into this. So if your fix works then let's go with that.

ghiscoding commented on slick.groupitemmetadataprovider.js:
```diff
@@ -20,9 +20,9 @@
+   * @param _options
    */
-  function GroupItemMetadataProvider(options) {
+  function GroupItemMetadataProvider(_options) {
```
Just a small change, I would prefer to see this one keep the previous naming since anything that starts with an underscore `_myVar` is typically a private variable, while this one is obviously a public variable.

ghiscoding replied:
If you don't want to conflict with the other variable, then maybe just name it `inputOptions` or anything else that make sense

arashdalir replied:
okay, productive feedback, will change the name to `inputOptions` and recommit.

arashdalir replied:
`_options` renamed to `inputOptions`

ghiscoding commented on slick.groupitemmetadataprovider.js:
```diff
@@ -42,8 +42,20 @@
+    {
+      options = $.extend(true, {}, _defaults, options, _options);
+
+      return this;
```
maybe remove the empty space and why are you returning `this` ? For the `this` we typically add a variable in the class as `var _self = this;` on top. Why not return the options instead of this?

arashdalir replied:
returning `this` would allow function chaining, but as the other functions of this class do not allow function chaining, I will remove the line altogether if that's okay. the function won't return anything.

ghiscoding replied:
ah ok yeah, function chaining is nice but we rarely use it, probably won't be used much in this context anyway so yeah removing it is probably better

arashdalir replied:
return value removed.

ghiscoding commented on slick.groupitemmetadataprovider.js:
```diff
@@ -42,8 +42,18 @@
       includeHeaderTotals: false
     };
 
-    options = $.extend(true, {}, _defaults, options);
```
now thinking about it, the fact that you removed this options copy will break most everyone's code. You have to set the options somewhere, so now unless the user is calling `setOptions` he won't have any options (not even the defaults). I suggest that you add this extend at least in the `init()` method.

arashdalir replied:
please check [line 47](https://github.com/6pac/SlickGrid/blob/778b1e6460f56ccda01fd7db07bda56ae66bf695/slick.groupitemmetadataprovider.js#L47). the function is called once per constructor, so no need to call it in init function...

ghiscoding replied:
ok I see but I would still prefer the changes that I mentioned for the `init` and `setOptions` which is what all other plugins/controls are doing. I don't think we should extend the defaults when calling `setOptions`.

arashdalir replied:
please check https://github.com/6pac/SlickGrid/pull/466#discussion_r382673606

ghiscoding commented on slick.groupitemmetadataprovider.js:
```diff
@@ -42,8 +42,18 @@
+
+    function setOptions(inputOptions)
+    {
+      options = $.extend(true, {}, _defaults, options, inputOptions);
```
considering what I wrote on the other extend that you removed, after putting it back in the `init()`, you should change this extend here with `options = $.extend({}, options, newOptions);` Same as what I did in the Grid Menu on this [line](https://github.com/6pac/SlickGrid/blob/master/controls/slick.gridmenu.js#L150), basically this should only change the options you are passing, nothing else and certainly not the defaults

arashdalir replied:
please check line 47 :) or the comment on your code review

arashdalir replied:
using `$.extend(true, {}, _defaults, ...)` ensures that the `_defaults` value is not changed (also show in following GIF.  
in fact, it doesn't change any of the objects other than the first (`target`) according to its [documentation](https://api.jquery.com/jquery.extend/)

`init` is not called on creation. as a result, the `options` will not be populated unless the plugin is initiated. the way I have called the function ensures the functionality is not changed in any way, as the execution path stays the same (extending the object using the `$.extend` function is done on virtually the same execution path location).  
I could easily change the execution path and call the `setOption` in `init` function, but cannot guaranty that the functionality doesn't break - in case `options` is used for any reason before plugin initialization.

![no-defaults-change](https://user-images.githubusercontent.com/2896170/75050962-4cb23e00-54cd-11ea-85a7-51bc41c9c087.gif)

ghiscoding replied:
the defaults are to be used only if you don't have anything set in your options and should be used only on first file load not afterward. What is wrong with the original code `options = $.extend(true, {}, _defaults, options);`? Does that give you issues?

arashdalir replied:
well, the way I have defined the function, the `_defaults` is practically only going to be used on first initiation. true, it will be used every time, but then overwritten by the values already set to `options`.  
the problem would be that the same code then would practically be used twice: once during the construction, once during the `setOptions`

ghiscoding replied:
The thing is that I would still prefer to see this `options = $.extend({}, options, newOptions);` in the `setOptions`, that is without the defaults

ghiscoding replied:
so like I said I would prefer if you do the changes mentioned, but if you want to stick with your code, at least put it on the same line so it's more obvious that it's called and set, at first glance I didn't see it at all.

arashdalir replied:
I would work on it on Monday.
